src/inmanta/server/services/compilerservice.py

```diff
@@ -218,7 +218,10 @@ async def run(self, force_update: Optional[bool] = False) -> Tuple[bool, Optiona
             if not repo_url:
                 if not os.path.exists(os.path.join(project_dir, "project.yml")):
                     await self._warning(f"Failed to compile: no project found in {project_dir} and no repository set")
-                await self._end_stage(0)
+                LOGGER.info("Installing missing modules")
```

Comment:
I think this should only run in case there is an update (i.e. force_update is true or the project is not on disk yet)

Reply:
I've made some changes for the other branch (`repo_url` is set). For this branch, I'm not sure how to implement this though. If I understand correctly `not repo_url` means that the server does not necessarily control the project directory (not sure about this). If that's true, then how do we know when we need to install modules?

Reply:
`not repo_url` means that if the project is not already on disk, the server won't be able to install it.

Reply:
i.e. it doesn't know the URL to clone  from

Reply:
So then how do we differentiate between the first compile (that needs to install modules) and later compiles?

Reply:
I think this logic would benefit from some restructuring.

Reply:
Doesn't `not repo_url` mean that a client-side compile was done instead of a server-side compile. If no `repo_url` is set, server-side compiles should be disabled entirely, no?

Reply:
I was inclined to agree with you but `tests/server/test_compilerservice.py::test_server_recompile` seems to contradict this. It sets up the project in the server's environment storage directory, then calls `client.notify_change()` to trigger a recompile.

Reply:
refactored as discussed (I think).

Reply:
thx